File: nodejs/roles.js
const Angle = require("./math/angle");
const Point = require("./math/point");

class BallFollower {
    applyOn(robot, snapshot) {
        if (snapshot.ball) {
            robot.moveToBall();
        } else {
            robot.moveToPoint(Point.ORIGIN);
        }
    }
}

class Goalkeeper {
    applyOn(robot, snapshot) {
        let ball = snapshot.ball ? snapshot.ball.position : Point.ORIGIN;
        let target = new Point(ball.x, -0.55);
        if (robot.position.dist(target) < 0.01) {
            robot.lookAtAngle(Angle.degrees(90));
        } else {
            robot.moveToPoint(target);
        }
    }
}

module.exports = {
    BallFollower: BallFollower,
    Goalkeeper: Goalkeeper
};

File: python/roles.py
from math_utils.point import Point
from math_utils.angle import degrees

class BallFollower:
    def applyOn(self, robot, snapshot):
        if snapshot.ball != None:
            robot.moveToBall()
        else:
            robot.moveToPoint(Point.ORIGIN)

class Goalkeeper:
    def applyOn(self, robot, snapshot):
        if snapshot.ball != None:
            ball = snapshot.ball.position
        else:
            ball = Point.ORIGIN

        target = Point(ball.x, - 0.55)
        if robot.getPosition().dist(target) < 0.01:
            robot.lookAtAngle(degrees(90))
        else:
            robot.moveToPoint(target)